tests/base.py

```diff
@@ -0,0 +1,305 @@
+        )
+
+        catalogs = menagerie.get_catalogs(conn_id)
+
+        # Ensure our selection affects the catalog
+        expected_selected = [tc.get('stream_name') for tc in test_catalogs]
+        for cat in catalogs:
+            catalog_entry = menagerie.get_annotated_schema(conn_id, cat['stream_id'])
+
+            # Verify all testable streams are selected
+            selected = catalog_entry.get('annotated-schema').get('selected')
+            print("Validating selection on {}: {}".format(cat['stream_name'], selected))
+            if cat['stream_name'] not in expected_selected:
+                self.assertFalse(selected, msg="Stream selected, but not testable.")
+                continue # Skip remaining assertions if we aren't selecting this stream
+            self.assertTrue(selected, msg="Stream not selected.")
+
+            if select_all_fields:
+                # Verify all fields within each selected stream are selected
+                for field, field_props in catalog_entry.get('annotated-schema').get('properties').items():
+                    field_selected = field_props.get('selected')
+                    print("\tValidating selection on {}.{}: {}".format(
+                        cat['stream_name'], field, field_selected))
+                    self.assertTrue(field_selected, msg="Field not selected.")
+            else:
+                # Verify only automatic fields are selected
+                expected_automatic_fields = self.expected_automatic_fields().get(cat['stream_name'])
+                selected_fields = self.get_selected_fields_from_metadata(catalog_entry['metadata'])
+                self.assertEqual(expected_automatic_fields, selected_fields)
+
+
+    def get_sync_start_time(self, stream, bookmark):
+        """
+        Calculates the sync start time, with respect to the lookback window
+        """
+        conversion_day = dt.now().replace(hour=0, minute=0, second=0, microsecond=0, tzinfo=None) - timedelta(days=self.lookback_window)
+        bookmark_datetime = dt.strptime(bookmark, self.BOOKMARK_FORMAT)
+        start_date_datetime = dt.strptime(self.start_date, self.START_DATE_FORMAT)
+        return  min(bookmark_datetime, max(start_date_datetime, conversion_day))
+
+
+    # TODO is this still useful now that we have get_stream_name?
+    def get_record_count_by_stream(self, record_count, stream):
+        count = record_count.get(stream)
+        if not count:
+            stream_name = self.custom_reports_names_to_ids().get(stream)
+            return record_count.get(stream_name)
+        return count
+
+
+    def get_bookmark_value(self, state, stream):
+        bookmark = state.get('bookmarks', {})
+        stream_bookmark = bookmark.get(stream)
+        stream_replication_key = self.expected_metadata().get(stream,set()).get('REPLICATION_KEYS')
+        if stream_bookmark:
+            return stream_bookmark.get(stream_replication_key)
+        return None
```

Comment:
Also copypasta from GA4 and may not be relevant, especially the function with an outdated todo

Reply:
Done.